blog/2024-02-27-how-to-keep-mqtt-data-python/index.md

```diff
@@ -0,0 +1,199 @@
+
+Let’s see how we store the data in the database:
+
+```python
+# Write received message to the bucket
+# by using topic as an entry name
+await bucket.write(message.topic, message.payload)
+```
+
+This was quite straightforward. We used the topic name as the entry name and wrote the MQTT payload there.
+
+## **Publishing Data to MQTT Topic**
+
+Upon launching the script, it remains idle since there's no data incoming from MQTT. To get it running, you need to publish some data. My preferred tool for this is **mosquitto_pub**. For those using Ubuntu, this tool comes with the `mosquitto-clients` package.
+
+```bash
+$ sudo apt-get install mosquitto-clients
+$ mosuitto_pub -t topic-1 -m "Hello, topic-1!"
+$ mosuitto_pub -t topic-2 -m "Hello, topic-2!"
+```
+
+## **Getting Data From ReductStore**
+
+You've learned to pull data from MQTT and save it to ReductStore. Now, we need a basic Python script to read this data from storage:
+
+```python
+import asyncio
+from reduct import Client, Bucket
+
+async def read_data():
+    async with Client("http://127.0.0.1:8383") as client:
+        bucket: Bucket = await client.get_bucket('mqtt')
+        for entry in await bucket.get_entry_list():
+            async for record in bucket.query(entry.name):
+                data = await record.read_all()
+                print(f"Received message {data} with ts={record.timestamp} from {entry.name} is read from the bucket")
+
+if __name__ == "__main__":
+    asyncio.run(read_data())
+```
+
+We have already covered how to create a client and acquire a bucket. Now, let's move on to data reading:
+
+```python
+for entry in await bucket.get_entry_list():
+    async for record in bucket.query(entry.name):
+        data = await record.read_all()
+        print(f"Received message {data} with ts={record.timestamp} from {entry.name} is read from the bucket")
+```
+
+As you can see, it is also very easy. We browse all entries in the `mqtt` bucket, query all records from each entry, and then print their timestamp and content. That's it.
+
+## Best Practices[](https://www.reduct.store/blog/tutorials/iot/how-to-keep-mqtt-data-node#best-practices)
+
+The provided example is basic and might not cover all the complexities you may face in a real-world application. Here are some tips to help you construct a strong and efficient IoT application using ReductStore and MQTT:
+
+- Create a ReductStore bucket with a FIFO quota to prevent disk overwriting in the future.
+- Use token authentication to protect your data. You can generate an access token using either the **[Web Console](https://github.com/reductstore/web-console)** or the **[CLI client](https://cli.reduct.store/)**.
+- Map MQTT5 properties to ReductStore labels. This will facilitate data filtering during querying or replication.
+- Use **[ReductCLI](http://cli.reduct.store/)** for data replication or backup purposes.
```

Comment:
link to blog/news/reductstore-8-released if they want to know more about replication